src/lib/Gen/adoc/AdocGenerator.cpp

```diff
@@ -105,9 +156,22 @@ buildOne(
     if(! errors.empty())
         return {errors};
 
-    SinglePageVisitor visitor(*ex, corpus, os);
+    std::string_view tagFilePath = corpus.config->tagfilePath;
+    if(! tagFilePath.empty())
+    {
+        if (auto err = files::createDirectory(files::getParentDir(tagFilePath))) 
+            return err;
+    }
+
+    auto tagFileWriterEx = createTagfileWriter(corpus, tagFilePath);
+    if(! tagFileWriterEx)
+        return tagFileWriterEx.error();
+
+    auto const justFileName = path::filename(anchorFileName);
+    SinglePageVisitor visitor(*ex, corpus, os, justFileName, std::move(*tagFileWriterEx));
```

Comment:
This unnecessary coupling between these visitors and the tagfile writer seems to be what's causing much of the problem because now you want to propagate these properties up to the API. But there's not needed there. `build` already has all the information it needs about the tagfile path from the corpus. 

Then it's easier to check if the tagfilePath is empty or not and run that step or not. If we even need a visitor for that (the visitor operates in the corpus but the domCorpus has the refs), the tagfile visitor could be much simpler because it would have the exact same logic for all nodes. 

In fact, if anything is wrong about `build` is that it should be simpler: not more complex. Because it also already has the information about the output path from the config options. In fact, I'm going to implement this right now in my branch.

Reply:
I understand your point about simplifying the `build` function and reducing the coupling between the visitors and the `TagfileWriter`. However, there's a key detail I’d like to clarify: the visitors manage specific information that the `TagfileWriter` relies on for proper tagfile generation.

- **MultiPageVisitor** determines the file names of the generated documentation files and passes them to the `TagfileWriter`.
- **SinglePageVisitor** provides unique data, such as the total number of pages, which the `TagfileWriter` uses for its logic.

This creates a necessary dependency between the visitors and the `TagfileWriter`.

Do you have any suggestions for how the `TagfileWriter` could receive this data without being directly coupled to the visitors?

Reply:
> MultiPageVisitor determines the file names of the generated documentation

That comes from the Dom. Not from the visitor.

A visitor has access to the build, which has access to the Dom, which has access to the xrefs (The code literally says `builder.domCorpus.getXref(I)`). 

The visitor has no information the generator doesn't. That's the information it has:

```
ExecutorGroup<Builder>& ex_;
std::string_view outputPath_;
Corpus const& corpus_;
```

The generator has direct access to the Dom without the Visitor or the Builder. A separate TagfileVisitor can take everything from the Dom directly without duplicated logic for single-page and multi-page visitors.

> SinglePageVisitor provides unique data, such as the total number of pages

If it's a single page, the number of pages is not very special information. The number of pages is 1. 

For multipage, it's the number of symbols + overloads (which will be simplified to the number of symbols in the future).

In any case, an XML writer should be able to write an array without knowing the number of elements in advance. Arrays are not prefixed by the number of elements in XML.

Reply:
If I recall correctly, we included page numbers in the anchor based on a discussion (possibly on Slack or in a previous PR review). This design made sense because SinglePageVisitor calculates this information, and the TagfileWriter uses it.

If your intention is to discard page number information entirely, the design of TagfileWriter and related code could indeed be simplified. 

Could you clarify if that’s the direction you’re suggesting?

Reply:
What is a page number in the anchor? I'm not following. Do you have an example in a tagfile where the page number is included?

I don't see page numbers in any of these files: https://github.com/boostorg/url/tree/develop/doc/tagfiles

I don't know where the page numbers would go or what that would mean because there is no order relationship between the pages in MrDocs.

All I see in the SinglePageVisitor is `page = numPages_`, which is the index of the current page so that pages are in order in the final output regardless of the process being asynchronous. But I don't see it being used in the output.